package/sbp_nmea2000_bridge/src/sbp_nmea2000_bridge.cpp

```diff
@@ -0,0 +1,917 @@
+        N2kMsg.AddByte(/*tracked=*/0xFF);
+      }
+      NMEA2000.SendMsg(N2kMsg);
+
+      d << "\tDone.\n";
+    }
+
+    void callback_sbp_heartbeat(u16 sender_id, u8 len, u8 msg[],
+                                void *context) {
+      UNUSED(sender_id);
+      UNUSED(len);
+      UNUSED(msg);
+      UNUSED(context);
+      d << __FUNCTION__ << "\n";
+
+      // This checks for new CAN messages and sends out a heartbeat.
+      NMEA2000.ParseMessages();
+
+      d << "\tDone.\n";
+    }
+}  // namespace
+
+int main(int argc, char *argv[]) {
+  logging_init(cProgramName);
+
+  if (parse_options(argc, argv) != 0) {
+    piksi_log(LOG_ERR, "Invalid arguments.");
+    usage(argv[0]);
+    exit(EXIT_FAILURE);
+  }
+
+  // Get the current state of the CAN interfaces.
+  int state_can0, state_can1;
+  piksi_check(can_get_state(cInterfaceNameCan0, &state_can0),
+              "Could not get %s state.", cInterfaceNameCan0);
+  piksi_check(can_get_state(cInterfaceNameCan1, &state_can1),
+              "Could not get %s state.", cInterfaceNameCan1);
+
+  // Bring the CAN interfaces down if they're up.
+  if(state_can0 != CAN_STATE_STOPPED){
+    piksi_check(can_do_stop(cInterfaceNameCan0),
+                "Could not stop %s.", cInterfaceNameCan0);
+  }
+  if(state_can1 != CAN_STATE_STOPPED){
+    piksi_check(can_do_stop(cInterfaceNameCan1),
+                "Could not stop %s.", cInterfaceNameCan1);
+  }
+
+  // Set the CAN bitrate. This must be done before turning them back on.
+  d << "Setting bitrate to " << bitrate_can0 << " and " << bitrate_can1 << "\n";
+  piksi_check(can_set_bitrate(cInterfaceNameCan0, bitrate_can0),
+              "Could not set bitrate %" PRId32 " on interface %s",
+              bitrate_can0, cInterfaceNameCan0);
+  piksi_check(can_set_bitrate(cInterfaceNameCan1, bitrate_can1),
+              "Could not set bitrate %" PRId32 " on interface %s",
+              bitrate_can1, cInterfaceNameCan1);
+
+  // Set the controlmode to enable flexible datarate (FD) frames.
+  // This is not supported yet in hardware, actually, but leave this commented
+  // code here for potential future use.
+  // This must be done before turning the interfaces on.
+//   can_ctrlmode ctrlmode_fd_on = {CAN_CTRLMODE_FD, CAN_CTRLMODE_FD};
+//   piksi_check(can_set_ctrlmode(cInterfaceNameCan0, &ctrlmode_fd_on),
+//               "Could not enable FD frames on %s.", cInterfaceNameCan0);
+//   piksi_check(can_set_ctrlmode(cInterfaceNameCan1, &ctrlmode_fd_on),
+//               "Could not enable FD frames on %s.", cInterfaceNameCan1);
+
+  // Turn the CAN interfaces on.
+  piksi_check(can_do_start(cInterfaceNameCan0),
+              "Could not start interface %s", cInterfaceNameCan0);
+  piksi_check(can_do_start(cInterfaceNameCan1),
+              "Could not start interface %s", cInterfaceNameCan1);
+
+  // Open CAN sockets.
+  int socket_can0 = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+  int socket_can1 = socket(PF_CAN, SOCK_RAW, CAN_RAW);
```

Comment:
~Suggestion: interleave`socket` call with the `if_nametoindex` call so it's clear that second call (with no changed parameters) is intentional.~ EDIT: suggest making a helper call for setting up each CAN socket, and bailing early if anything fails int he setup, a lot of duplicated code in the setup process.